src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.Http2KeepAlivePing.cs

```diff
@@ -0,0 +1,371 @@
+                    // Simulate inactive period:
+                    await Task.Delay(5_000);
+
+                    // We may receive one RTT PING in response to HEADERS.
+                    // Upon that, we expect to receive at least 1 keep alive PING:
+                    Assert.True(_pingCounter > 1);
+                }
+                else
+                {
+                    // We should receive no more KeepAlive PINGs
+                    Assert.True(_pingCounter <= 1);
+                }
+
+                await TerminateLoopbackConnectionAsync();
+
+                List<Frame> unexpectedFrames = new List<Frame>();
+                while (_framesChannel.Reader.Count > 0)
+                {
+                    Frame unexpectedFrame = await _framesChannel.Reader.ReadAsync();
+                    unexpectedFrames.Add(unexpectedFrame);
+                }
+
+                Assert.False(unexpectedFrames.Any(), "Received unexpected frames: \n" + string.Join('\n', unexpectedFrames.Select(f => f.ToString()).ToArray()));
+            }, NoAutoPingResponseHttp2Options).WaitAsync(TestTimeout);
+        }
+
+        [OuterLoop("Runs long")]
+        [Fact]
+        public async Task KeepAliveConfigured_NoPingResponseDuringActiveStream_RequestShouldFail()
+        {
+            await Http2LoopbackServer.CreateClientAndServerAsync(async uri =>
+            {
+                SocketsHttpHandler handler = new SocketsHttpHandler()
+                {
+                    KeepAlivePingTimeout = TimeSpan.FromSeconds(1.5),
+                    KeepAlivePingPolicy = HttpKeepAlivePingPolicy.WithActiveRequests,
+                    KeepAlivePingDelay = TimeSpan.FromSeconds(1)
+                };
+                handler.SslOptions.RemoteCertificateValidationCallback = delegate { return true; };
+
+                using HttpClient client = new HttpClient(handler);
+                client.DefaultRequestVersion = HttpVersion.Version20;
+
+                // Warmup request to create connection:
+                HttpResponseMessage response0 = await client.GetAsync(uri);
+                Assert.Equal(HttpStatusCode.OK, response0.StatusCode);
+
+                // Actual request:
+                await Assert.ThrowsAsync<HttpRequestException>(() => client.GetAsync(uri));
+
+                // Let connection live until server finishes:
+                await _serverFinished.Task;
+            },
+            async server =>
+            {
+                await EstablishConnectionAsync(server);
+
+                // Warmup the connection.
+                int streamId1 = await ReadRequestHeaderAsync();
+                await GuardConnctionWriteAsync(() => _connection.SendDefaultResponseAsync(streamId1));
+
+                // Request under the test scope.
+                int streamId2 = await ReadRequestHeaderAsync();
+
+                DisablePingResponse();
+
+                // Simulate inactive period:
+                await Task.Delay(6_000);
```

Comment:
Do we really need such long delay here? Is there any way to implement this on synchronization primitives like AutoResetEvent, ManualResetEvent, etc?

Reply:
> Do we really need such long delay here? 

By having a big difference between the `KeepAlivePingDelay` and making sure the tests runs sequentially, we can be *relatively* safe from CI timing issues. Removing any of these conditions will lead to certain failures IMO.

Note that he minimum value for `KeepAlivePingDelay` and `KeepAlivePingTimeout`  is 1 sec, so I can't make this faster by using low delay values

>  Is there any way to implement this on synchronization primitives

I don't see how. The API exposes `TimeSpan` parameters without any testability-focused abstractions that would allow us to use events.

I understand that the extra ~1 minute added here is very bad from testing perspective, but the only alternative I see is to remove the KeepAlive tests, which means loosing coverage. I wish we used whitebox testing techniques in dotnet/runtime ...

Reply:
Ok. I got it.